src/main/java/com/werfen/report/model/GridReportDataSource.java

```diff
@@ -11,35 +10,40 @@
 
 public class GridReportDataSource extends JRAbstractBeanDataSource {
 
-    private final GridReportData rows;
+    private final GridPageDataSource dataSource;
     private Iterator<GridReportRow> rowIterator;
     private GridReportRow currentRow;
 
-    public GridReportDataSource(GridReportData rows) {
+    public GridReportDataSource(GridPageDataSource dataSource) {
         super(true);
-        this.rows = rows;
-        if (nonNull(rows)) {
-            this.rowIterator = this.rows.getRows().listIterator();
+        this.dataSource = dataSource;
+        if (this.dataSource.getRowCount() > 0) {
```

Comment:
After assigning the data source, the code should be the same as in moveFirst() method implementation. In other words, when constructing a GridReportDataSource, we are automatically moving to first row.  Thus, I would suggest calling this method instead of implementing something here directly (which right now is different from what we have in moveFirst()).